pkg/agent/bgp/gobgp/gobgp.go

```diff
@@ -0,0 +1,288 @@
+	}
+	if peerAddress == "" {
+		return nil, fmt.Errorf("BGP peer IP address must be specified")
+	}
+
+	var routes []bgp.Route
+	fn := func(destination *gobgpapi.Destination) {
+		if destination == nil {
+			return
+		}
+		routes = append(routes, bgp.Route{Prefix: destination.GetPrefix()})
+	}
+
+	request := &gobgpapi.ListPathRequest{
+		TableType: gobgpTableType,
+		Family:    &gobgpapi.Family{Afi: toGoBGPFamilyAfi(net.IsIPv6String(peerAddress)), Safi: gobgpapi.Family_SAFI_UNICAST},
+		Name:      peerAddress,
+	}
+
+	if err := s.server.ListPath(context.TODO(), request, fn); err != nil {
+		return nil, err
+	}
+	return routes, nil
+}
+
+func toGoBGPPath(route *bgp.Route) *gobgpapi.Path {
+	isIPv6 := net.IsIPv6CIDRString(route.Prefix)
+	goBGPIPFamily := toGoBGPFamilyAfi(isIPv6)
+	prefix, _ := netip.ParsePrefix(route.Prefix)
+	nlri, _ := anypb.New(&gobgpapi.IPAddressPrefix{
+		Prefix:    prefix.Addr().String(),
+		PrefixLen: uint32(prefix.Bits()),
+	})
+
+	var attrs []*anypb.Any
+	a1, _ := anypb.New(&gobgpapi.OriginAttribute{
+		Origin: 0,
+	})
+	var a2 *anypb.Any
+	if isIPv6 {
+		a2, _ = anypb.New(&gobgpapi.MpReachNLRIAttribute{
+			Family:   &gobgpapi.Family{Afi: goBGPIPFamily, Safi: gobgpapi.Family_SAFI_UNICAST},
+			NextHops: []string{ipv6AllZero},
+			Nlris:    []*anypb.Any{nlri},
+		})
+	} else {
+		a2, _ = anypb.New(&gobgpapi.NextHopAttribute{
+			NextHop: ipv4AllZero,
+		})
+	}
+	attrs = append(attrs, a1, a2)
+
+	return &gobgpapi.Path{
+		Family: &gobgpapi.Family{Afi: goBGPIPFamily, Safi: gobgpapi.Family_SAFI_UNICAST},
+		Nlri:   nlri,
+		Pattrs: attrs,
+	}
+}
+
+func toGoBGPFamilyAfi(isIPv6 bool) gobgpapi.Family_Afi {
+	if isIPv6 {
+		return gobgpapi.Family_AFI_IP6
+	}
+	return gobgpapi.Family_AFI_IP
+}
+
+func toGoBGPPeer(peerConfig bgp.PeerConfig) *gobgpapi.Peer {
+	peer := &gobgpapi.Peer{
+		Conf: &gobgpapi.PeerConf{
+			NeighborAddress: peerConfig.Address,
+			PeerAsn:         uint32(peerConfig.ASN),
+			AuthPassword:    peerConfig.Password,
+		},
+		AfiSafis: []*gobgpapi.AfiSafi{
+			{
+				Config: &gobgpapi.AfiSafiConfig{
+					Family: &gobgpapi.Family{
+						Afi:  toGoBGPFamilyAfi(net.IsIPv6String(peerConfig.Address)),
+						Safi: gobgpapi.Family_SAFI_UNICAST,
+					},
+					Enabled: true,
+				},
+				MpGracefulRestart: &gobgpapi.MpGracefulRestart{
+					Config: &gobgpapi.MpGracefulRestartConfig{
+						Enabled: true,
+					},
+				},
+			},
+		},
+	}
+	if peerConfig.Port != nil {
+		peer.Transport = &gobgpapi.Transport{
+			RemotePort: uint32(*peerConfig.Port),
+		}
+	}
+	if peerConfig.MultihopTTL != nil && *peerConfig.MultihopTTL != bgp.DefaultBGPMultihopTTL {
+		peer.EbgpMultihop = &gobgpapi.EbgpMultihop{
+			Enabled:     true,
+			MultihopTtl: uint32(*peerConfig.MultihopTTL),
+		}
+	}
+	if peerConfig.GracefulRestartTimeSeconds != nil {
+		peer.GracefulRestart = &gobgpapi.GracefulRestart{
+			Enabled:     true,
+			RestartTime: uint32(*peerConfig.GracefulRestartTimeSeconds),
+		}
+	}
```

Comment:
could you clarify why `MultihopTTL` and `GracefulRestartTimeSeconds` are treated differently. In particular, we only set the `MultihopTTL` if it's not set to the default value, but we don't do the same thing for `GracefulRestartTimeSeconds`. Same question for `GetPeers` above.

My opinion is that we should only check for `nil`, not for the default. There is no harm in setting it explicitly to the default, and it makes the code easier to read / maintain.

Reply:
> could you clarify why MultihopTTL and GracefulRestartTimeSeconds are treated differently. In particular, we only set the MultihopTTL if it's not set to the default value, but we don't do the same thing for GracefulRestartTimeSeconds. Same question for GetPeers above.

My bad. They should be treated in the same way. The `MultihopTTL` and `GracefulRestartTimeSeconds` are always set since they have default values. Not sure if we should do the nil checks for these two fields, but it is harmless to do so. It seems that it is a convention for pointer field to do nil check even it has a default value (I saw such usage in other projects).